### PR TITLE
[processing] Tests for eliminate

### PR DESCRIPTION
**Pull request by bstroebl**
an new input layer is necessary for two reasons:
1) has to have a polygon that can be eliminated (i.e. that is adjacent to at least one existing polygon)
2) has to be in an editable format (shp used in this case)

### CODE REVIEW

**Comment by m-kuhn** [comment]
Nice,

Are you able to use one of the existing source datasets instead?

**Comment by bstroebl** [comment]
> Are you able to use one of the existing source datasets instead?

no, for the resasons given above

**Comment by m-kuhn** [comment]
> 2) has to be in an editable format (shp used in this case)

Does that mean that the layer will be modified?

**Comment by bstroebl** [comment]
> Does that mean that the layer will be modified?

it is modified during the algorithm but rolled back later on. I originally wrote this to eliminate within an existing layer (as an addition to fTools). Would need reimplementation to do it otherwise (although this would make more sense from the point of processing).

**Comment by m-kuhn** [comment]
Are you referring to the `rollBack()` on the almost-last line? That's just executed in memory (edit buffer), nothing ever gets written to the file, so it's never actually modified.

Concerning the other question, the main reason for asking for using default files is, that it also tests for null geometries and attributes and the geometries in there have simple coordinates which makes it easier to interpret results.

I wonder if it makes sense to put additional source files into a subfolder, so future test writers use the test files which are present on toplevel as source file. I'd really like to prevent a mess like we have in the qgis testdata folder.

**Comment by bstroebl** [comment]
> Are you referring to the rollBack() on the almost-last line? That's just executed in memory (edit buffer), nothing ever gets written to the file, so it's never actually modified.

could be deleted then, couldn't it?

> Concerning the other question, the main reason for asking for using default files is, that it also tests for null geometries and attributes and the geometries in there have simple coordinates which makes it easier to interpret results.

I did not want to change the existing source file because I do not know where it is used already. All I did is add one additional data set (snapped, so coordinates should be simple, too). 

> I wonder if it makes sense to put additional source files into a subfolder, so future test writers use the test files which are present on toplevel as source file. I'd really like to prevent a mess like we have in the qgis testdata folder.

Basically one should check if an existing dataset can be used for the intended test before creating a new one. Another way would be to provide test files that include more situations like (in this case) adjacent polygons, overlapping polygons, polygons completely inside others etc. Didn't see a null geometry in polys.gml BTW

**Comment by m-kuhn** [comment]
> Basically one should check if an existing dataset can be used for the intended test before creating a new one.

Yes, but if there are more than just a few datasets in the toplevel directory that won't happen.

> Another way would be to provide test files that include more situations like (in this case) adjacent polygons, overlapping polygons, polygons completely inside others etc. Didn't see a null geometry in polys.gml BTW

In this case I would retire the current one (by either putting it into a subfolder or removing it totally and updating existing tests). I think we'll never get a "one-suits-all" file, but it can surely be better than at the moment.

So the options are:

1. Create a new polygon file - building on the existing one - with a broader applicability
2. Move the additional file required for this test into a subfolder

**Comment by bstroebl** [comment]
test failed  :-(
I was not sure if I am allowed to edit the existing dataset (although editing gml files is not possible anyways, maybe we should switch to shp for test input?)
how if any test writer may add (only add) features to existing input files? If he does he would have to rerun all testd that use this input. Or if each test always brings its own input(s) (strict naming convention necessary)?

**Comment by m-kuhn** [comment]
> I was not sure if I am allowed to edit the existing dataset (although editing gml files is not possible anyways, maybe we should switch to shp for test input?)

No you are not. But you are welcome to add an updated and improved copy.
The test suite is still small enough that we can handle such a change. But in the future this will probably be no longer an easy task.

Why not keep a few generic test datasets that can be used for 70% of all cases and let tests that don't fall into this category drop their required files into a subfolder where we don't care if it's messy?

**Comment by bstroebl** [comment]
> Why not keep a few generic test datasets that can be used for 70% of all cases and let tests that don't fall into this category drop their required files into a subfolder where we don't care if it's messy?

:+1: 
but how would you create generic test datasets? Can only be done if the existing one is continually improved by test authors of further tests.
What do you think about shp so editing is easier?

**Comment by m-kuhn** [comment]
> but how would you create generic test datasets? Can only be done if the existing one is continually improved by test authors of further tests.

do the best that can be done now and then leave it and if it's not compatible with future tests, drop a new file in the subfolder. So no continuous updating. But if you think that they can be improved with...

> adjacent polygons, overlapping polygons, polygons completely inside others etc. Didn't see a null geometry in polys.gml 

... please update the current dataset now, it would be great. If not, please put your source datafile into a subfolder.

Shapefiles are really a relict of the last century with attribute name restrictions, no support for mixed geometries...

**Comment by bstroebl** [comment]
as I wrote, I have to use an editable file format. If I use gml as input I get an error in the log because line 299 fails:
`if inLayer.changeGeometry(mergeWithFid, newGeom):`
gml files cannot be edited in QGIS so they may not be the best choice for input in this case. What format would you suggest instead of shp?

**Comment by m-kuhn** [comment]
Oh, now I see. The file doesn't actually get changed (only its edit buffer), but despite this fact, it cannot be put into edit mode and therefore the algorithm fails. To me this looks more like a problem of the algorithm (it's not applicable to a read-only file although it does not require to update the file) and a feature of the dataset to highlight such problems. The source files should never be changed throughout the test process (and if we ever require to do this, we will add an explicit option to take a copy of it before).

Bottomline: fix the algorithm or put the shapefile, spatialite file or some other writeable file into a subfolder.

**Comment by bstroebl** [comment]
The source file polys.gml has two problems:

1. it is not editable; fixing the algorithm would mean reimplementing but would of course be the best option
2. there is no feature in it that can be eliminated (i.e. merged with an adjacent feature)

> Bottomline: fix the algorithm or put the shapefile, spatialite file or some other writeable file into a subfolder.

For the time beeing I would add the shp in a subfolder (any preference for the name of it?) `special_input`?

**Comment by bstroebl** [comment]
but the test fails anyways because the output seems to be `MULTILINESTRING` instead of polygon. I wonder why it runs in QGIS, though.

**Comment by m-kuhn** [comment]
> it is not editable; fixing the algorithm would mean reimplementing but would of course be the best option

as I said, that's not a problem of the source file. If you need such a source file, put it into a separate folder.

> there is no feature in it that can be eliminated (i.e. merged with an adjacent feature)

For that, it would be great if you could replace the current source file, but since it already fails on condition 1, I'm not sure if you want to do that :)

**Comment by m-kuhn** [comment]
@bstroebl I fixed some debug output problems and it now fails with

> | Error in selection input: Cannot convert ""polys.5"" to float

**Comment by bstroebl** [comment]
@m-kuhn This is one of the errors I throw; the alg checks for the field type with `type()` and tries to convert the input value (which always is a string) into this type. no idea, why it thinks `fid` is of type float...

File: python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml [hunk 1]
```diff
@@ -122,3 +122,57 @@ tests:
...
+    results:
+      OUTPUT:
+        name: expected/eliminate_largest_area.gml
+        type: vector
+
+  # case 2: merge with smallest area
+  - algorithm: qgis:eliminatesliverpolygons
+    name: Eliminate sliver polygons smallest area
+    params:
+      ATTRIBUTE: '"fid"'
```
Comment (m-kuhn) [inline python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml]: Can you replace this with

    ATTRIBUTE: 'fid'

and

      COMPARISONVALUE: 'polys.5'

I'll fix the test creator script accordingly

**Comment by bstroebl** [comment]
I tried here in QGIS 2.14.0 with polys.shp
`inLayer.dataProvider().fields()[0].type()` returns `10`. Any idea why the test returns `6`?
BTW I am examinig if I can change the code to support gml as input.

**Comment by m-kuhn** [comment]
When I checked before, it was looking for a field called `"fid"` instead of `fid` and got -1 as field index and that somehow messed things up...

**Comment by m-kuhn** [comment]
Manually merged

Thanks again for your patience.

Remaining issue was that there was still a `'"fid"'` and `'"polys.5"'` present

The test generator has been improved so that should not happen again

**Comment by bstroebl** [comment]
ok, will check if field exists in next version

**Comment by bstroebl** [comment]
Mathias could you reopen this PR? otherwise I would create a new one.

**Comment by m-kuhn** [comment]
This pull request has been merged. It's ok to create a new one.
